File: solutions/36.cpp
#include <array>
#include <vector>

class Solution {
 public:
  bool isValidSudoku(std::vector<std::vector<char> >& board) {
    std::array<std::array<unsigned, 9>, 9> rows{{{0}}}, cols{{{0}}},
        tiles{{{0}}};

    for (size_t row = 0; row < board.size(); ++row) {
      for (size_t col = 0; col < board.front().size(); ++col) {
        size_t tile = ((row / 3) * 3) + (col / 3);
        if (board[row][col] == '.') {
          continue;
        }
        unsigned value = static_cast<unsigned>(board[row][col] - '0' - 1);
        if (rows[row][value] or cols[col][value] or tiles[tile][value]) {
          return false;
        }
        ++rows[row][value];
        ++cols[col][value];
        ++tiles[tile][value];
      }
    }
    return true;
  }
};

File: solutions/36.py
#!/usr/bin/env python3

import numpy as np
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def contains_duplicated_numbers(e) -> bool:
            e = e[e != '.']
            return len(e) != len(set(e))

        board = np.array(board)
        for index in range(9):
            row = board[index, :]
            if contains_duplicated_numbers(row):
                return False
            col = board[:, index]
            if contains_duplicated_numbers(col):
                return False
        for r, c in zip(range(3), range(3)):
            quad = board[3 * r:3 * r+3, 3 * c:3 * c+3]
            if contains_duplicated_numbers(quad):
                return False
        return True
